### DISCUSSION TITLE
What is a Prompt Theme Engine, and what should it do?

### DISCUSSION THREAD

**Discussion by Rycieos** [liquidprompt/liquidprompt#786] [Ideas]
### What is a Prompt Theme Engine?
I think a distinction can be made between a "prompt theme" and a "prompt theme engine" (from here on shortened to PTE).

In the far past, users would always manually customize their `PS1`s. Eventually, some had the brilliant idea to make it more complicated, and then to share that creation with others, @nojhan included. Themes, by definition, are opinionated, and are not for everyone. Usually that meant a user would find the theme that fit them best, but eventually this lead to customization preferences in themes, and eventually the framework to drop in a theme over top of the theme, and which point the base theme was not really a theme anymore, but a true PTE.

There are a number of problems with this current state, which I will define in a section below.

But to answer the question, a PTE exists to **interface between the terminal and the running theme, and to expose contextual data for the theme to use**.

### The problems with current PTEs
These problems are mostly universal, Liquid Prompt included.
1. Due to PTEs starting as simply a theme, there are still a lot of embedded opinions, leaving customization lacking.
2. Every PTE needs to handle the data gathering from scratch. This leaves many PTEs with subpar data quality.
3. Similarly, many PTEs lack data types or terminal integration features that a user might want.
4. Software developers that are good at writing a theme are usually not also good at writing good, performant data gathering code.
5. A PTE is often written for one or two shells, leaving other users out of luck.

### A universal PTE
To solve these problems, I am envisioning a future with one PTE. This "universal" PTE would handle all of the terminal and contextual data problems for themes, and let theme creators just worry about the theme part. It would have zero opinions about what the prompt looks like. And it would be lightning fast.

### Problems to solve
There are obviously a lot of issues to figure out first, or this probably would have been done already. Here is what I am stuck on yet:

#### What things should be configured in a PTE vs in the theme?
Some things are somewhat obvious. A specific feature (say, Git integration) should probably be enabled/disabled at the PTE level. But some things are not obvious. What mark to display for a Git branch depends on where a theme puts it, and under what conditions it will show it. Should a threshold for Git data be configured at the PTE? What if the logic for the threshold is in the hands of the theme? So if you leave it up to the theme to define, what if every theme defines the same logic, but slightly differently, leaving people confused?

Most PTEs available today lean very heavily toward putting everything in the PTE. Liquid Prompt, Starship, and Oh My Posh are all the most like a true PTE (in my opinion), and yet all have very centralized configs. They all make it difficult for a theme to completely change everything about how a prompt is displayed. (For example, I tried to recreate Liquid Prompt's default theme in Starship, and was only able to get about 75% of the way there.)

#### How to handle theme customization?
One of the most often requested features in Liquid Prompt is custom functions for overriding a specific part of the theme. Any PTE worth anything should have a way for a user to extend a theme to customize parts of it, but to reuse the parts they do not want to change.

Ideally this could extend to distributed themes, that could depend on a different theme to extend it.

#### How to handle theme distribution?
I would like the install process for popular themes to remain the same. Which likely rules out "install <universal PTE>, then run `pte install my-theme`.

#### How to handle data gathering?
Clearly the PTE does the actually data gathering. But the question of "when" and "who controls it" is much harder. If we gather data when the theme wants it (like most current PTEs do), we miss out on possible speedups from parallelization. But if we gather ALL data in parallel right away, we risk gathering more data than the theme will actually need, which could perform worse. Ideally we could predict what data the theme will ask for, and gather it early to have it ready.

#### How to support plugins?
Another one of the most requested features in Liquid Prompt is plugin support for custom data sources. Ideally the universal PTE supports all data types, but since that likely is not possible, supporting at least a rudimentary plugin system would allow for themes or users to use data that is not so standard.

#### Can data truly be unopinionated?
The problem with current PTEs is that they are too opinionated about what the prompt will look like. But is it possible to provide structured data to themes without being opinionated at all? If not, I don't see such a universal PTE ever becoming popular.

### Prior art
Clearly anything I reason about will be influenced by Liquid Prompt. But this has also been greatly inspired by [Starship](https://starship.rs/) and [Powerlevel10k](https://github.com/romkatv/powerlevel10k)'s [gitstatus](https://github.com/romkatv/gitstatus)

### Work so far
This has been mostly nothing but a dream of mine for a few years, and a more concrete idea for the past year. I created [MOST](https://github.com/Rycieos/most) (stands for **M**ulti-shell **O**pen-source **S**criptable **T**erminal Prompt Engine) about 6 months ago as PoC (and also as a way to learn more Rust), and have not touched it since. It is very rough, and not at all ready for use yet. I think more thought and design needs to go into it before I can continue, which is why I am posting this.

Its biggest features are:
1. A daemon, which results in the fastest prompt generation I have ever seen.
2. Embedded scripting engine, which gives total control over execution to the theme.

**Comment by nojhan** [comment]
I'm writing this post to put some thoughts there, but those are not well structured.

I feel that there are in fact three layers in such an approach:
1. the data-gathering daemon,
2. the theme engine,
3. the actual theme.

# Data daemon

This may make use of existing libraries that access the system states, like the ones that would been made by any software like [htop](https://htop.dev/) or [glances](https://nicolargo.github.io/glances/).

But, it also needs to be lightning fast when anything changes. If I remember well the last time I read Gitstatusd's code, the daemon triggers a state check every time it is queried. But I think we can do even better with a daemon, if we register directories of interest and use the filesystem to monitor them (like what [watchdog](https://github.com/gorakhargosh/watchdog) is doing, worked well for me on [flicksave](https://github.com/nojhan/flicksave)).

The data daemon strikes me as being easily generic, with the design bias being that it would group data by features as seen by a prompt, and carry only semantics, which should not be a big bias.

API design rules, as I see it, would be:
- the data daemon API holds entry points taking an (optional) directory and returning a data structure,
- nothing in the returned data is about the presentation, so there should not be any icons or colors, just names, and semantic tags.
- all the data should be returned, so no threshold, and no filtering of any sort,
- it should provide the list of possible states, with their corresponding semantics:
    - qualitative states (e.g. text of graphical capabilities),
    - quantitative states:
        - numerical range (e.g. Git's number of commits),
        - categories of increasing "importance" (e.g. Git's status).


# Theme engine

The theme engine is more biased, given its targeted prompt application, of course. But One can probably find a design that's pretty much generic among the existing themes. LP has a nice diversity of them, so we can have a nice view of the current state of the art.

Here are the features I noted I would have needed while implementing themes:
- the theme engine serves "segments", a string of characters that can be interpreted by a text interface (probably the terminal, but it may be a status line as well. I think it should be restricted to text UI, though, and not go as far as [conky](https://github.com/brndnmtthws/conky), which targets GUI, but also provides a ─clunky─ TUI backend.
- the theme engine should hold any information that may be useful for the presentation, without concern about what is actually displayed by the actual theme:
    - a unique name/ID for the segment,
    - a mapping of data states to semantic states,
        - simple state => style (e.g. "more than 5 subshells" == "alert style"),
        - ranges for mapping numerical states to a semantic one (e.g. "more than 50 lines altered" == "git is dirty")
    - a style is a list of (non-mutually exclusive) display atoms:
        - text (useful for accessibility, e.g. display Git's state explicitly as text),
        - icons,
        - colors,
        - styles (italic, bold, etc.),
        - URLs.
    - explicit starting and ending characters/styles (which are important in TUIs) for each segment.

I feel that the theme engine may live as a daemon, alongside the data daemon.
In its simplest form, it may very well just be a pre-configured dictionary, mapping from the features proposed (declared?) by the data daemon. But I think it is probable that a procedural implementation would be needed at some point to be truly generic.


# Actual theme

The theme itself decides what to display, and holds the enable/disable configuration.
It should probably declare what features it is interested in at startup time, so that the daemons can disable any useless feature, to avoid useless computation.

It would probably be better implemented in shell script (?)

**Reply by Rycieos** [reply]
Thanks for taking the time to think about these things and to write them all down. I took some time myself to think about it.

I agree that there would need to be 3 layers. I imagined that the data and theme engine layers would be distinct, but would likely live inside the same daemon process. However, you got me thinking. Let me lay out the problem as I see it first.

> The theme itself ... holds the enable/disable configuration.

This is the problem. Imagine a Git data source. There might be a config option similar to `LP_ENABLE_GIT`, which would enable or disable the Git feature as a whole. This should, obviously to me, live inside the data layer. But now imagine a setting that controls whether information on Git remotes is shown. It could live in the data layer, except that a theme might need that data for reasons other than displaying it directly (for example, only show branch information if there is an upstream remote set). So it could live in the theme layer, except that many themes might want to have a setting like this, and they might name it different things. Now if a user switches themes, a feature they thought they enabled will be disabled again. Hopefully that explains the problem well.

What if the theme engine was swappable? Then these config options could live in that layer, and any theme built on that theme engine would use the same config options. This would also give much more flexibility to themes without giving them so much power that they need to do everything themselves. The theme engine would make lots of assumptions (use segments or not, group certain things) which gives themes less to do.

The question then becomes how are theme engines built? Are they compiled plugins for the data daemon, or are they written in scripting language like the themes? My *MOST* proof of concept had the theme engine baked in to the daemon (therefore written in Rust), but it did almost no hand-holding, so most of the prompt logic was in the theme. Themes were written in [Rune](https://github.com/rune-rs/rune), a somewhat Rust-like scripting language.

I both love and hate the idea of making the theme engines swappable. Figuring out how to distribute themes was going to be difficult enough, but also needing to distribute theme engines would be even harder.

This "config problem" is the largest problem, in my opinion. Users want to try different themes (or even use multiple themes at the same time) without needing to figure out how to translate their configuration file. We need to support a theme having its own configuration options for very specific things, but also be able to hook into config options that are more "global" without actually having an effect on the data layer. I have thought about this for years now and have not thought of a good solution.
